### PR TITLE
perf(model): Add xFormers in GPTJ

### LINKED ISSUE #73: [Super AI] [Good First Issue] Convert GPTJ Self Attention to XFormer
Description:
[GPTJ](https://huggingface.co/EleutherAI/gpt-j-6b) is one of the [most popular](https://huggingface.co/models?pipeline_tag=text-generation) Open Sources of GPT model available on the Internet. However, the current Huggingface [code](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py) of GPTJ is using the old Attention mechanism. In this task, we want to convert the original code from GPTJ to support Flash Attetnion by doing

1. Download the model from Huggingface and try to run inference using their official implementation code. (6.7B Version is enough)
2. Convert to both:
- Pytorch 2.0 Attention: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
- XFormer : https://github.com/HazyResearch/flash-attention
3. Try to run inference again and get the same result (logits)

Deliverables:
- Convert/Inference Code Script
- PR
- Unit Test to make sure converted model have similar output to the original one

### PR DESCRIPTION
**Pull request by kriangkraitan**
Add xFormers in GPTJ
Copied code from [https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py](url)
1. Custom `_attn ()` function in Class `GPTJAttention(nn.Module)` to use xFormers ---> `_attn_xformers()`
2. Add `GPTJConfig(PretrainedConfig)` to be able to switch between self-attention mode and using xFormers mode.
3. Add gptj_test to test output when using xFormers.

Close #73

### CODE REVIEW

File: test_py.py
Comment (boat1603) [inline test_py.py:1]: ???

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:1]: I think you can import some original function from `transformers`. You can see my example in the [Electra repo](https://github.com/OpenThaiGPT/Pretraining-ELECTRA/blob/main/src/electra_pretraining/electra/flash_attn_model.py). You can inherit some module to reduce number of lines of code.

**Review by boat1603 — COMMENTED** [review]
You should `pull main` to update your branch up to date.

File: test_py.py
Comment (kriangkraitan) [reply test_py.py:1]: It has already been deleted.

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py
Comment (kriangkraitan) [reply src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:1]: I changed class `GPTJConfig` from my code. and replaced by import the originals functions from the transformer and fix running `GPTJConfig` in `gptj_test.py` file

Comment (kriangkraitan) [reply src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:1]: and use funtion(`create_sinusoidal_positions,get_embed_positions,apply_rotary_pos_emb`,) from Transformers library

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py [hunk 1]
```diff
@@ -0,0 +1,808 @@
+# coding=utf-8
```
[the remark this answers is not in the document]
Comment (kriangkraitan) [reply src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py]: I replace the entire model file with import from `transfomers` library and define function `get_output()` to get output and swith between originals_attention and using xformers

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import torch
+from transformers.models.gptj.modeling_gptj import (
+    GPTJAttention,
+    GPTJModel,
+)
+from transformers import AutoTokenizer
+from xformers.components.attention.core import scaled_query_key_softmax, bmm
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:7]: Use [memory efficient attention](https://facebookresearch.github.io/xformers/components/ops.html) instead

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        query,
+        key,
+        value,
+        xops.LowerTriangularMask(),
+    )
+
+    return attn_output, None
+
+
+def get_output(pretrained_name, use_xformers, input_text):
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:39]: get output should be moved to test file

File: src/data/pyproject.toml [hunk 4]
```diff
@@ -14,5 +14,17 @@ classifiers = [
     "Programming Language :: Python :: 3",
 ]
 dependencies = [
+<<<<<<< HEAD
```
Comment (new5558) [inline src/data/pyproject.toml:17]: What is this? Please fix

File: src/data/pyproject.toml [hunk 5]
```diff
@@ -14,5 +14,17 @@ classifiers = [
...
+    "xformers>=0.0.18",
+    "llama@git+https://github.com/facebookresearch/llama.git",
+    "sentencepiece>=0.1.98",
+    "fairscale>=0.4.13",
+    "lion-pytorch>=0.0.7",
+    "transformers>=4.28.1",
+    "datasets>=2.12.0",
+    "zstandard>=0.21.0"
+]
+=======
```
Comment (new5558) [inline src/data/pyproject.toml:27]: Fix

Comment (new5558) [reply src/data/pyproject.toml:27]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py (hunk 3 above)
Comment (kriangkraitan) [reply src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py]: done

File: src/data/pyproject.toml [hunk 6]
```diff
@@ -14,5 +14,14 @@ classifiers = [
...
 ]
 dependencies = [
+    "xformers>=0.0.18",
+    "llama@git+https://github.com/facebookresearch/llama.git",
+    "sentencepiece>=0.1.98",
+    "fairscale>=0.4.13",
+    "lion-pytorch>=0.0.7",
+    "transformers>=4.28.1",
+    "datasets>=2.12.0",
+    "zstandard>=0.21.0"
```
Comment (new5558) [inline src/data/pyproject.toml:24]:
```suggestion
    "zstandard>=0.21.0",
```

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    head_mask=None,
+):
+    if attention_mask is not None:
+        raise TypeError("Not support manual attention mask")
+
+    if head_mask is not None:
+        raise TypeError("Not support head_mask")
+
+    # Attention output
+    attn_output = xops.memory_efficient_attention_forward(
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:48]: Why `memory_efficient_attention_forward` not `memory_efficient_attention`

File: src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    # Mask heads if we want to
+    if head_mask is not None:
+        attn_weights = attn_weights * head_mask
+    # attn_output = bmm(attn_weights, value)
+    attn_output = bmm(attn_weights, value)
+
+    return attn_output, attn_weights
+
+
+def _attn_xformers(
```
[the remark this answers is not in the document]
Comment (kriangkraitan) [reply src/model/openthaigpt_pretraining_model/gptj/gptj_model_xformers.py:33]: done

**Review by new5558 — APPROVED** [review]
